daemon/containerd/image_delete.go

```diff
@@ -78,9 +140,179 @@ func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force,
+		return true
+	}
+
+	if i := strings.IndexRune(imageID, ':'); i >= 0 {
+		return strings.HasPrefix(imageID[i+1:], possiblePrefix)
+	}
+
+	return false
+}
+
+// isSingleReference returns true if there is no other image
+// with the same target, not including dangling images
+func (i *ImageService) isSingleReference(ctx context.Context, imgID image.ID) bool {
+	imgs, err := i.client.ImageService().List(ctx, "target.digest=="+imgID.String())
+	if err != nil {
+		return true
+	}
+	var notDangling int
+	for _, img := range imgs {
+		if !isDanglingImage(img) {
+			notDangling++
+		}
+	}
+	return notDangling == 1
+}
+
+// ImageDeleteConflict holds a soft or hard conflict and an associated error.
+// Implements the error interface.
+type imageDeleteConflict struct {
+	hard    bool
+	used    bool
+	imgID   image.ID
+	message string
+}
+
+func (idc *imageDeleteConflict) Error() string {
+	var forceMsg string
+	if idc.hard {
+		forceMsg = "cannot be forced"
+	} else {
+		forceMsg = "must be forced"
+	}
+
+	return fmt.Sprintf("conflict: unable to delete %s (%s) - %s", stringid.TruncateID(idc.imgID.String()), forceMsg, idc.message)
+}
+
+func (idc *imageDeleteConflict) Conflict() {}
+
+// imageDeleteHelper attempts to delete the given image from this daemon. If
+// the image has any hard delete conflicts (child images or running containers
+// using the image) then it cannot be deleted. If the image has any soft delete
+// conflicts (any tags/digests referencing the image or any stopped container
+// using the image) then it can only be deleted if force is true. If the delete
+// succeeds and prune is true, the parent images are also deleted if they do
+// not have any soft or hard delete conflicts themselves. Any deleted images
+// and untagged references are appended to the given records. If any error or
+// conflict is encountered, it will be returned immediately without deleting
+// the image. If quiet is true, any encountered conflicts will be ignored and
+// the function will return nil immediately without deleting the image.
+func (i *ImageService) imageDeleteHelper(ctx context.Context, imgID image.ID, records *[]types.ImageDeleteResponseItem, force, prune, quiet bool) error {
+	// First, determine if this image has any conflicts. Ignore soft conflicts
+	// if force is true.
+	c := conflictHard
+	if !force {
+		c |= conflictSoft
+	}
+
+	if conflict := i.checkImageDeleteConflict(ctx, imgID, c); conflict != nil {
+		image, err := i.resolveImage(ctx, imgID.String())
+		if err != nil {
+			return err
+		}
+		if quiet && (!isDanglingImage(image) || conflict.used) {
+			// Ignore conflicts UNLESS the image is "dangling" or not being used in
+			// which case we want the user to know.
+			return nil
+		}
+
+		// There was a conflict and it's either a hard conflict OR we are not
+		// forcing deletion on soft conflicts.
+		return conflict
+	}
+
+	parent, err := i.Parent(ctx, imgID)
+	if err != nil {
+		parent = ""
+	}
```

Comment:
This is mostly a straight copy-paste of the old code in order to get some comments, and as @vvoland mentioned it might not even make sense to prune parents with the containerd store. If we do keep it, yeah this should be looked at.